### PR TITLE
Implement deletion of all consents on tenant deletion

### LINKED ISSUE #174: Implement deletion of all consents of a given tenant
**Description:**
Implement the methods to delete all Purposes, Purposes Categories, PII Categories, and Receipts of a given tenant.

This is required in the tenant deletion process.

**Related Issues:**
wso2/carbon-identity-framework#2914

### PR DESCRIPTION
**Pull request by sumedhe**
## Purpose
Implement the methods to delete all Purposes, Purposes Categories, PII Categories, and Receipts of a given tenant.

Closes #174

### CODE REVIEW

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/dao/ReceiptDAO.java [hunk 1]
```diff
@@ -70,6 +70,16 @@
      */
     void deleteReceipt(String receiptID) throws ConsentManagementException;
 
+    /**
+     * Delete all {@link Receipt} of a given principal tenant id.
+     *
+     * @param tenantId Id of the tenant
+     * @throws ConsentManagementException
+     */
+    default void deleteReceiptsByPrincipalTenantId(int tenantId) throws ConsentManagementException {
```
Comment (omindu) [inline components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/dao/ReceiptDAO.java:79]: We might need to delete the receipts by service provide tenant ID as well. Please check

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java [hunk 2]
```diff
@@ -671,9 +738,31 @@ public void deleteReceipt(String receiptId) throws ConsentManagementException {
...
+     *
+     * @param tenantId Id of the tenant.
+     * @throws ConsentManagementException Consent Management Exception.
+     */
+    @Override
+    public void deleteReceipts(int tenantId) throws ConsentManagementException {
+
+        if (tenantId <= 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("Tenant id is not found in the request or invalid tenant id");
```
Comment (omindu) [inline components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java:752]: We can just say, invalid tenant id: {tenantId}

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java [hunk 3]
```diff
@@ -671,9 +738,31 @@ public void deleteReceipt(String receiptId) throws ConsentManagementException {
...
+            }
+            throw handleClientException(ERROR_CODE_TENANT_ID_REQUIRED, null);
+        }
+
+        getReceiptsDAO(receiptDAOs).deleteReceiptsByPrincipalTenantId(tenantId);
```
Comment (omindu) [inline components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java:757]: same as above. make sure to indicate the tenant Id in the exception

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java (hunk 2 above)
Comment (omindu) [reply components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java:752]: Fix in other applicable places

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java [hunk 4]
```diff
@@ -671,9 +738,31 @@ public void deleteReceipt(String receiptId) throws ConsentManagementException {
...
+     * @throws ConsentManagementException Consent Management Exception.
+     */
+    @Override
+    public void deleteReceipts(int tenantId) throws ConsentManagementException {
+
+        if (tenantId <= 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("Tenant id is not found in the request or invalid tenant id");
+            }
+            throw handleClientException(ERROR_CODE_TENANT_ID_REQUIRED, null);
```
Comment (omindu) [inline components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java:754]: same as above. make sure to indicate the tenant Id in the exception. same as above. make sure to indicate the tenant Id in the exception

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/dao/ReceiptDAO.java (hunk 1 above)
Comment (sumedhe) [reply components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/dao/ReceiptDAO.java:79]: Will check. THX!

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java (hunk 2 above)
Comment (sumedhe) [reply components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java:752]: 👍

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java (hunk 4 above)
Comment (sumedhe) [reply components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/ConsentManagerImpl.java:754]: 👍

**Comment by omindu** [comment]
Please add unit tests for new functions

File: components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/constant/ConsentConstants.java [hunk 5]
```diff
@@ -203,7 +204,19 @@
...
+
+        ERROR_CODE_DELETE_PURPOSE_CATEGORIES_BY_TENANT_ID("CM_00092", "Error occurred while deleting " +
+                "purpose categories from DB for the tenant: %s."),
+        ERROR_CODE_DELETE_PII_CATEGORIES_BY_TENANT_ID("CM_00093", "Error occurred while deleting " +
+                "PII categories from DB for the tenant: %s."),
+        ERROR_CODE_DELETE_PURPOSES_BY_TENANT_ID("CM_00094", "Error occurred while deleting purpose " +
+                "from DB for the tenant: %s."),
+        ERROR_CODE_DELETE_RECEIPTS_BY_PRINCIPAL_TENANT_ID("CM_00095", "Error occurred while deleting " +
+                "receipt from DB for the tenant: %s."),
+        ERROR_CODE_DELETE_SP_ASSOC_BY_SP_TENANT_ID("CM_00095", "Error occurred while deleting " +
```
Comment (omindu) [inline components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/constant/ConsentConstants.java:218]: Same error code repeated

Comment (sumedhe) [reply components/org.wso2.carbon.consent.mgt.core/src/main/java/org/wso2/carbon/consent/mgt/core/constant/ConsentConstants.java:218]: Fixed in #179

**Pull request merged**